File: backend/app/Services/Coupons/CouponQuote.php
<?php

namespace App\Services\Coupons;

use App\Models\Coupon;

class CouponQuote
{
    public function __construct(
        public readonly Coupon $coupon,
        public readonly float $eligibleSubtotal,
        public readonly float $discount,
        public readonly bool $freeShipping,
        public readonly string $message,
    ) {}

    public function toArray(float $subtotal, float $shippingFee, float $total): array
    {
        return [
            'code' => $this->coupon->code,
            'title' => $this->coupon->title ?: $this->coupon->code,
            'type' => $this->coupon->type->value,
            'type_label' => $this->coupon->type->label(),
            'message' => $this->message,
            'eligible_subtotal' => round($this->eligibleSubtotal, 2),
            'discount_amount' => round($this->discount, 2),
            'free_shipping' => $this->freeShipping,
            'subtotal' => round($subtotal, 2),
            'shipping_fee' => round($shippingFee, 2),
            'total' => round($total, 2),
        ];
    }
}

File: mobile/lib/features/shop/data/models/coupon_quote.dart
class CouponQuote {
  final String code;
  final String title;
  final String type;
  final String message;
  final double eligibleSubtotal;
  final double discountAmount;
  final bool freeShipping;
  final double subtotal;
  final double shippingFee;
  final double total;
  final bool hasFreeShipping;
  final String? deliveryLabel;

  const CouponQuote({
    required this.code,
    required this.title,
    required this.type,
    required this.message,
    required this.eligibleSubtotal,
    required this.discountAmount,
    required this.freeShipping,
    required this.subtotal,
    required this.shippingFee,
    required this.total,
    required this.hasFreeShipping,
    this.deliveryLabel,
  });

  factory CouponQuote.fromJson(Map<String, dynamic> json) {
    final delivery = json['delivery'];
    return CouponQuote(
      code: (json['code'] as String?) ?? '',
      title: (json['title'] as String?) ?? '',
      type: (json['type'] as String?) ?? '',
      message: (json['message'] as String?) ?? '',
      eligibleSubtotal: (json['eligible_subtotal'] as num?)?.toDouble() ?? 0,
      discountAmount: (json['discount_amount'] as num?)?.toDouble() ?? 0,
      freeShipping: json['free_shipping'] as bool? ?? false,
      subtotal: (json['subtotal'] as num?)?.toDouble() ?? 0,
      shippingFee: (json['shipping_fee'] as num?)?.toDouble() ?? 0,
      total: (json['total'] as num?)?.toDouble() ?? 0,
      hasFreeShipping: json['has_free_shipping'] as bool? ?? false,
      deliveryLabel: delivery is Map ? delivery['label'] as String? : null,
    );
  }
}
